File: notes/amazon/valid_parenthesis.js
var isValid = function (s) {

  const stack = new Array();
  const open = new Set(["(", "{", "["]);
  const close = new Set([")", "}", "]"]);
  const map = { ")": "(", "}": "{", "]": "[" }

  for (let i = 0; i < s.length; i++) {
    let chr = s[i]
    if (close.has(chr)) {
      const popped = stack.pop();
      if (popped !== map[chr]) {
        return false;
      }
    } else {
      if (!open.has(chr)) {
        return false;
      }
      stack.push(chr)
    }
  }
  if (stack.length === 0) {
    return true;
  } else {
    return false;
  }
};

File: 08_leet_code/daily/validParenthesis.java
class Solution {
    public boolean isValid(String s) {
        Map <String, String> openToClose = new HashMap<String, String>();
        Map <String, String> closeToOpen = new HashMap<String, String>();
        Stack<String> stack = new Stack<String>();

        openToClose.put("(", ")");
        openToClose.put("[", "]");
        openToClose.put("{", "}");
        closeToOpen.put(")", "(");
        closeToOpen.put("]", "[");
        closeToOpen.put("}", "{");
        
        for (int i = 0; i < s.length(); i++) {
            char chr = s.charAt(i);
            String str = Character.toString(chr);
        
            if (openToClose.get(str) instanceof String) {
                stack.push(str);
            } else {
                String top = "";
                if (stack.size() > 0) {
                    top = stack.peek();
                }
                String shouldBeTop = closeToOpen.get(str);
                if (!top.equals(shouldBeTop)) {
                    return false;
                } else {
                    stack.pop();
                }
            }
        }
        
        if (stack.size() > 0) {
            return false;
        } else {
            return true;
        }
    }
}